# .01 Importing libraries

In [1]:
# Importing libraries
import pandas as pd
import numpy as ny
import os

In [ ]:
# Importing orders.csv data as df_ords

In [61]:
df_ords = pd.read_csv(r'D:\Rich\Bus Analyst\29-06-2023 Instacart Basket Analysis\02 Data\Original Data\orders.csv',index_col=False)

In [62]:
path = r'D:\Rich\Bus Analyst\29-06-2023 Instacart Basket Analysis'

In [4]:
path

'D:\\Rich\\Bus Analyst\\29-06-2023 Instacart Basket Analysis'

In [63]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

# Step 2 Variable that doesn't need to be included in analysis

In [9]:
# Importing products.csv data as df_prods

In [64]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'),index_col = False)

In [11]:
# Exercise 4.4 Data wrangling Drop function in pd

In [65]:
# Dropping eval_set column from orders.csv
df_ords = df_ords.drop(columns = ['eval_set'])

In [8]:
# Using the value_counts() function on "days_since_prior_order"
df_ords['days_since_prior_order'].value_counts(dropna=False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

# Step 3 Renaming columns

In [66]:
# Renaming 'order-dow' to 'orders_day_of_week'
df_ords.rename(columns = {'order_dow': 'orders_day_of_week'},inplace = True)

In [ ]:
# Print first 5 rows with new column name 'orders_day_of_week'

In [67]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [ ]:
# Changing a variable

In [12]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [13]:
# Using describe() to change the 'order_id' from a number to a string
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [14]:
df_ords['order_id'].dtype

dtype('O')

In [ ]:
# Importing departments data set

In [15]:
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'),index_col = False)

In [16]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [17]:
# Transposing departments data from one row and 22 columns to one column and 22 rows
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [18]:
# Creating a new dataframe to a transposed version calling it "df_dep_t"
df_dep_t = df_dep.T

In [19]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [20]:
# Removing the "O" and turn first row of the dataframe into headers
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [21]:
# First step row of df_dep_t for the header
new_header = df_dep_t.iloc[0]

In [22]:
new_header

0    department
Name: department_id, dtype: object

In [23]:
df_dep_t_new = df_dep_t[1:]

In [24]:
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [25]:
# Take the data under the header row for a new df
df_dep_t_new = df_dep_t[1:]

In [26]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [26]:
# set the header row as the df header
df_dep_t_new.columns = new_header

In [27]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [28]:
# Data Dictionary using department data
data_dict = df_dep_t_new.to_dict('index')

In [29]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [30]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [31]:
# Print data_dict to determine waht department_id 19 means
print(data_dict.get('19'))

{'department': 'snacks'}


In [32]:
# Creating a subset for df_prods
df_snacks = df_prods[df_prods['department_id']==19]

In [33]:
df_prods['department_id']==19

0         True
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

In [34]:
# Indexing code within 'department_id' of 19
df_prods[df_prods['department_id']==19]

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [35]:
# Save results as df_snacks
df_snacks =df_prods[df_prods['department_id']==19]

In [65]:
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

In [37]:
# Another way to achieve results using 'loc' function
df_snacks_2 = df_prods.loc[df_prods['department_id']==19]

In [83]:
df_snacks_2

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [38]:
df_snacks_3 = df_prods.loc[df_prods['department_id'].isin([19])]

In [39]:
df_snacks_3

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [40]:
df_snacks_3 = df_prods.loc[df_prods['department_id'].isin([17,18,19])]

In [41]:
df_snacks_3

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
13             14                 Fresh Scent Dishwasher Cleaner        74   
14             15                       Overnight Diapers Size 6        56   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
...           ...                                            ...       ...   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49672       49668                 Apple Cinnamon Scented Candles       101   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
13                17     6.5  
14                18    11.2  
15                19     5.2  
24                19     1.9  
...              ...     ...  
49669             19     6.9  
49670             19     5.2  
49672             17     5.6  
49675             19     3.0  
49683             19     6.0  

[10430 rows x 5 columns]

In [39]:
# Exporting dataframes syntax
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders wranglers.csv'))

# Step 4 Busiest hours for placing orders

In [42]:
4.4 #4 Busiest hours for placing orders. Find the frequency of the corresponding variable 
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329        1             1                   2                  8   
1  2398795        1             2                   3                  7   
2   473747        1             3                   3                 12   
3  2254736        1             4                   4                  7   
4   431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [43]:
df_ords.sort_values('order_number', ascending=False)

order_id  user_id  order_number  orders_day_of_week  \
129907   1539024     7832           100                   3   
1828766  1305301   109846           100                   3   
725940   2916106    43682           100                   5   
2343405  1378313   141069           100                   5   
2628558  1223488   158262           100                   2   
...          ...      ...           ...                 ...   
2882084   931757   173912             1                   1   
929490   2222962    55745             1                   1   
2882075  1606642   173911             1                   4   
2882071   306688   173910             1                   6   
0        2539329        1             1                   2   

         order_hour_of_day  days_since_prior_order  
129907                   5                     6.0  
1828766                 10                     1.0  
725940                   9                     3.0  
2343405                  8                     7.0  
2628558                  9                     3.0  
...                    ...                     ...  
2882084                 10                     NaN  
929490                  15                     NaN  
2882075                 12                     NaN  
2882071                 18                     NaN  
0                        8                     NaN  

[3421083 rows x 6 columns]

# Step 4 Answer: I ran df_ords.sort_values('order_number',ascending = False) to get the highest number of orders and there are five with 100 orders and the hours of day are 5,10,9,8,and 9.

# Step 5 Determine the meaning behind value of 4 in 'department_id'

In [44]:
4.4 #5 Determine the meaning behind a value of "4" in the "department_id" column with the df_prods df using the data dictionary.
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [45]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

# Step 5 Answer is the meaning behind value 4 in the 'department_id' using the data dictionary is 'produce'

# Step 6 Sales team wants to know more about breakfast item sales

In [46]:
df_breakfast = df_prods[df_prods['department_id']==14]

In [47]:
df_prods['department_id']==14

0        False
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

In [48]:
df_prods[df_prods['department_id']==14]

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

In [ ]:
4.4 #6 Answer is creating a subset of breakfast items using the data ditctionary breakfast value is 14 and created a subset
        # df_prods[df_prods['department_id']==14] to create a list of breakfast items.     

# Step 7 Sales team wants to see products for dinner parties. Create a subset

In [49]:
df_prods[df_prods['department_id']==5]

product_id                               product_name  aisle_id  \
51             52                        Mirabelle Brut Rose       134   
118           119                     Chardonnay Paso Robles        62   
149           150                                  Brut Rosé       134   
233           234                          Tennessee Whiskey       124   
248           249             Pinot Grigio, California, 2010        62   
...           ...                                        ...       ...   
49548       49544                    Cabernet Sauvignon Wine        28   
49566       49562              Blanc De Noirs Sparkling Wine       134   
49610       49606                        Organic Natural Red        28   
49665       49661                                      Porto       134   
49688       49684  Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
51                 5    14.4  
118                5     5.5  
149                5    12.9  
233                5     3.1  
248                5     2.7  
...              ...     ...  
49548              5     9.8  
49566              5     5.7  
49610              5     6.8  
49665              5     8.2  
49688              5     5.3  

[1056 rows x 5 columns]

In [50]:
df_prods[df_prods['department_id']==7]

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
19             20  Pomegranate Cranberry & Aloe Vera Enrich Drink        98   
...           ...                                             ...       ...   
49659       49655                                     Apple Cider        98   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
19                 7     6.0  
...              ...     ...  
49659              7    10.7  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  

[4365 rows x 5 columns]

In [51]:
df_prods[df_prods['department_id']==12]

product_id                                       product_name  \
16             17                                  Rendered Duck Fat   
22             23                             Organic Turkey Burgers   
34             35     Italian Herb Porcini Mushrooms Chicken Sausage   
39             40  Beef Hot Links Beef Smoked Sausage With Chile ...   
83             84                                         Lamb Shank   
...           ...                                                ...   
49425       49421                                   Salame, Peppered   
49440       49436                              Imitation Crab Flakes   
49509       49505                                Hot Italian Sausage   
49655       49651                                       Beef Brisket   
49674       49670                                Beef Summer Sausage   

       aisle_id  department_id  prices  
16           35             12    17.1  
22           49             12     8.2  
34          106             12    15.1  
39          106             12    22.5  
83            7             12    24.3  
...         ...            ...     ...  
49425       106             12    12.2  
49440        15             12    23.5  
49509       106             12    15.8  
49655       122             12    20.7  
49674       106             12    19.2  

[907 rows x 5 columns]

In [52]:
df_prods[df_prods['department_id']==20]

product_id                                   product_name  aisle_id  \
48             49  Vegetarian Grain Meat Sausages Italian - 4 CT        14   
61             62        Premium Deli Oven Roasted Turkey Breast        96   
73             74                Artisan Chick'n & Apple Sausage        14   
84             85                           Soppressata Piccante        96   
108           109                         Grape Leaf Hummus Wrap        13   
...           ...                                            ...       ...   
49558       49554                          Roasted Garlic Hommus        67   
49564       49560    Selects Natural Slow Roasted Chicken Breast        96   
49585       49581                   Pinto Bean and Cheese Pupusa        13   
49609       49605                     Classic Hummus Family Size        67   
49642       49638           Organic Superfirm Vacuum Packed Tofu        14   

       department_id  prices  
48                20    10.1  
61                20    14.6  
73                20     3.0  
84                20     8.9  
108               20     7.1  
...              ...     ...  
49558             20    14.8  
49564             20    14.5  
49585             20    10.5  
49609             20     3.5  
49642             20     9.0  

[1322 rows x 5 columns]

In [53]:
df_din_party = df_prods.loc[df_prods['department_id'].isin([5,7,12,20])]

In [54]:
df_din_party

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

# Step 8 How many rows does the last DF have

The answer for the number of rows in the last DF is 7,650 

# Step 9 Data engineer team spotted something strange 'user_id' of '1'

In [74]:
df_1 = df_ords[df_ords['user_id']==1]

In [75]:
df_1

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   1187899        1            11                   4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

# Step 9 Anwer: I did a subset in 'user_id' for '1' with df_1 = df_ords[df_ords['user_id']==1] and retrieved '1' information.

# Step 10 Provide some details about the user's behavior. What basic stats can you provide?

# Step 10 Answer: Basic stats 'user_id' '1' has ordered 11 times, usually the first 4 days of the week, most of the orders are morning with four mid-day on. Had two orders the same day within an hour of each order. 

# Step 11 Structure and organization of notebook with headings and code comments

# Step 12 Export df_ords dataframe as 'orders_wrangled.csv' in 'Prepared Data' folder

In [89]:
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders wranglers.csv'))

# Step 13 Export df_dep_t_new dataframe as 'department_wrangled.csv' in 'Prepared Data' folder

In [90]:
df_dep.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'df_dep_t_new.csv'))

# Step 14 Save Jupyter notebook and submit for review